File: menubar/Sources/Pulse/CodexModel.swift
import Foundation

/// Best-effort current Codex model, read from the local Codex session logs. Menubar-only;
/// mirrors `Model.swift` for Claude. Every step is lenient — failure means "no model", never
/// an error, because the model is optional in the UI.

/// Find the model of the *last* `turn_context` record in a Codex rollout log, scanning from
/// the end. `turn_context` carries the model actually used for that turn, so a mid-session
/// model switch is reflected; `session_meta` / `thread_settings_applied` are ignored.
func extractLastCodexModel(_ content: String) -> String? {
    let lines = content.split(separator: "\n", omittingEmptySubsequences: false)
    for line in lines.reversed() {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        if trimmed.isEmpty { continue }
        guard let data = trimmed.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              obj["type"] as? String == "turn_context",
              let payload = obj["payload"] as? [String: Any],
              let model = payload["model"] as? String, !model.isEmpty else {
            continue
        }
        return model
    }
    return nil
}

/// Map a model slug to its display name via Codex's `models_cache.json`
/// (`models[].{slug, display_name}`). Any decode problem or unknown slug → the slug itself.
func codexModelDisplayName(_ slug: String, cache: Data?) -> String {
    guard let cache,
          let obj = try? JSONSerialization.jsonObject(with: cache) as? [String: Any],
          let models = obj["models"] as? [[String: Any]] else {
        return slug
    }
    for model in models where model["slug"] as? String == slug {
        if let name = model["display_name"] as? String, !name.isEmpty { return name }
    }
    return slug
}

/// Read the model from the most recently modified rollout log under `CODEX_HOME/sessions`.
func readCurrentCodexModel() -> CurrentModel? {
    let fm = FileManager.default
    let home = codexHome()
    let root = home.appendingPathComponent("sessions")
    guard let enumerator = fm.enumerator(
        at: root, includingPropertiesForKeys: [.contentModificationDateKey], options: [.skipsHiddenFiles]
    ) else {
        return nil
    }

    var best: URL?
    var bestDate = Date.distantPast
    for case let url as URL in enumerator where url.pathExtension == "jsonl" {
        let date = (try? url.resourceValues(forKeys: [.contentModificationDateKey]))?
            .contentModificationDate ?? Date.distantPast
        if date > bestDate {
            bestDate = date
            best = url
        }
    }

    guard let path = best,
          let content = try? String(contentsOf: path, encoding: .utf8),
          let slug = extractLastCodexModel(content) else {
        return nil
    }
    let cache = try? Data(contentsOf: home.appendingPathComponent("models_cache.json"))
    return CurrentModel(id: slug, name: codexModelDisplayName(slug, cache: cache))
}

File: tray-go/codex_model.go
package main

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
)

// readCurrentCodexModel reads the last turn_context model from the newest Codex rollout log.
// It is best-effort: Codex usage remains available when no local session log exists.
func readCurrentCodexModel() (*currentModel, error) {
	root := filepath.Join(codexHome(), "sessions")
	var best string
	var bestMod int64 = -1
	err := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
		if err != nil || d.IsDir() || filepath.Ext(path) != ".jsonl" {
			return nil
		}
		if info, e := d.Info(); e == nil && info.ModTime().UnixNano() > bestMod {
			best, bestMod = path, info.ModTime().UnixNano()
		}
		return nil
	})
	if err != nil || best == "" {
		return nil, os.ErrNotExist
	}
	b, err := os.ReadFile(best)
	if err != nil {
		return nil, err
	}
	id := extractLastCodexModel(string(b))
	if id == "" {
		return nil, os.ErrNotExist
	}
	return &currentModel{ID: id, Name: codexModelDisplayName(id)}, nil
}

func extractLastCodexModel(content string) string {
	lines := strings.Split(content, "\n")
	for i := len(lines) - 1; i >= 0; i-- {
		var record struct {
			Type    string `json:"type"`
			Payload struct {
				Model string `json:"model"`
			} `json:"payload"`
		}
		if json.Unmarshal([]byte(strings.TrimSpace(lines[i])), &record) == nil && record.Type == "turn_context" && record.Payload.Model != "" {
			return record.Payload.Model
		}
	}
	return ""
}

func codexModelDisplayName(slug string) string {
	b, err := os.ReadFile(filepath.Join(codexHome(), "models_cache.json"))
	if err != nil {
		return slug
	}
	var cache struct {
		Models []struct {
			Slug        string `json:"slug"`
			DisplayName string `json:"display_name"`
		} `json:"models"`
	}
	if json.Unmarshal(b, &cache) != nil {
		return slug
	}
	for _, model := range cache.Models {
		if model.Slug == slug && model.DisplayName != "" {
			return model.DisplayName
		}
	}
	return slug
}
